auth_fetch.go

```diff
@@ -0,0 +1,95 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/bitrise-io/go-utils/log"
+	"github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver/devportalservice"
+)
+
+// AppleAuth contains either Apple ID or APIKey auth info
+type AppleAuth struct {
+	AppleID *AppleIDAuth
+	APIKey  *devportalservice.JWTConnection
+}
+
+// AppleIDAuth contains Apple ID auth info
+type AppleIDAuth struct {
+	// Either session (if using 2FA) or username+password is required
+	session            string
+	username, password string
+
+	// Needed to deploy IPA to App Store Connect with altool (optional)
+	appSpecificPassword string
+}
+
+// MissingAuthConfigError is returned in case no usable Apple App Store Connect / Developer Portal authenticaion is found
+type MissingAuthConfigError struct {
+}
+
+func (*MissingAuthConfigError) Error() string {
+	return "Apple authentication not configured"
+}
+
+// FetchAppleAuthData return valid Apple ID or API Key based authentication data, from the provided Bitrise Service or manual inputs
+func FetchAppleAuthData(authSources []AppleAuthSource, inputs AppleAuthInputs) (AppleAuth, error) {
+	if err := inputs.Validate(); err != nil {
```

Comment:
it seems this line makes apple service auth related inputs required.
in case connection enabled (via `connection` input) and available than auth inputs are not required

Reply:
Indeed. Fixed.